csrc/host_ir/executor.cpp

```diff
@@ -7,13 +7,157 @@
 // clang-format on
 
 #include <dynamic_transform.h>
+#include <fusion_profiler.h>
 #include <host_ir/executor.h>
+#include <instrumentation.h>
 #include <ir/utils.h>
+#include <multidevice/communication.h>
+#include <multidevice/lower_communication.h>
+#include <multidevice/utils.h>
+#include <options.h>
+#include <runtime/allocations.h>
+#include <runtime/executor_dispatch.h>
 #include <runtime/executor_kernel_arg.h>
 #include <runtime/fusion_kernel_runtime.h>
 
 namespace nvfuser {
 
+HostIrExecutor::HostIrExecutor(
+    int64_t fusion_id,
+    int64_t concrete_id,
+    int64_t runtime_id,
+    int64_t group_id)
+    : ExecutorAbstract(fusion_id, concrete_id, runtime_id, group_id),
+      communicator_(&Communicator::getInstance()) {}
+
+bool HostIrExecutor::supported(Fusion* fusion) {
+  FUSER_PERF_SCOPE("HostIrExecutor::supported");
+  if (fusion->isA<hir::HostIrContainer>()) {
+    return true;
+  }
+  std::vector<Expr*> exprs = fusion->exprs();
+  if (std::any_of(exprs.begin(), exprs.end(), [](Expr* e) {
+        return isResharding(e) && isLowerableToCommunication(e);
+      })) {
+    NVF_ERROR(
+        std::all_of(
+            exprs.begin(),
+            exprs.end(),
+            [](Expr* e) {
+              return isResharding(e) && isLowerableToCommunication(e);
+            }),
+        "Could not execute fusion as all expressions in a host IR container must be communication based at this point.");
+    return true;
+  }
+  return false;
+}
+
+void HostIrExecutor::compile(Fusion* fusion) {
+  FUSER_PERF_SCOPE("HostIrExecutor::compile");
+  NVF_ERROR(
+      supported(fusion),
+      "HostIrExecutor does not support the Fusion provided.");
+  if (isProfilerEnabled()) {
+    FusionProfiler::segment(group_id_).startCompile();
+  }
+
+  if (fusion->isA<hir::HostIrContainer>()) {
+    host_ir_container_ = std::make_unique<hir::HostIrContainer>();
+    IrCloner cloner = Fusion::copy(fusion, host_ir_container_.get());
+    for (auto expr : fusion->as<hir::HostIrContainer>()->topLevelExprs()) {
+      host_ir_container_->pushBackTopLevelExprs(cloner.clone(expr));
+    }
+  } else {
+    std::vector<Expr*> exprs = fusion->exprs();
+    host_ir_container_ = std::make_unique<hir::HostIrContainer>();
+    IrCloner cloner = Fusion::copy(fusion, host_ir_container_.get());
+    for (Expr* e : exprs) {
+      std::vector<Communication*> communications =
+          lowerCommunication(cloner.clone(e));
+      for (auto* communication : communications) {
+        host_ir_container_->pushBackTopLevelExprs(communication);
+      }
+    }
+  }
+
+  if (isProfilerEnabled()) {
+    FusionProfiler::segment(group_id_).stopCompile();
+  }
+}
+
+bool HostIrExecutor::isCompiled() const {
+  return host_ir_container_ != nullptr;
+}
+
+namespace {
+// Host IR specific function, returns the at:Tensor (ordered list) associated
```

Comment:
In general, I found that passing the tensors through `unordered_map<Val*, at::Tensor>` is way more powerful and straightforward than dealing with `std::vector<at::Tensor>`, because with the latter we always need to find the relative index and loose track of intermediate tensors. In addition, I don't see any benefit in keeping an order in the inputs.

A map would allow to get rid of this type of helper function.

Also, I think the inputs, outputs and intermediate tensors should undergo an uniform treatment inside the Executor. Ideally, we want to allow the user to provide some preallocated output or intermediate buffer, or at least to handle allocations of those buffers independently of the execution.

Reply:
This follows the explicit contract with user's/integration. This contact is followed by KernelExecutor, FusionExecutorCache, and KernelRuntime, so I'm trying to stay consistent.

Reply:
I agree for inputs/outputs, but why would we do so for intermediate buffers? Who would provide those allocations?

Reply:
> This follows the explicit contract with user's/integration. This contact is followed by KernelExecutor, FusionExecutorCache, and KernelRuntime, so I'm trying to stay consistent.

This contract makes things much harder than they should in my opinion. What I am proposing is to change this contract to pass `unordered_map<Val*, at::Tensor>` instead of `std::vector<at::Tensor>`. I see no advantage of `std::vector` version, where we constantly need to retrieve the corresponding Val associated with the concrete tensor. 
This being said, nothing prevents us to have an overloaded user facing entrypoint taking `std::vector`, that would only be thin wrapper around the one taking `unorderd_map`.  

> I agree for inputs/outputs, but why would we do so for intermediate buffers? Who would provide those allocations?

The motivation behind this proposal is to have a uniform and more flexible treatment of allocation and pre-allocated buffers accross all level of execution:
- We might want to only pre-allocate some outputs or intermediate buffers and not all of them, and there is no way to indicate this when passing `std::vector<at::Tensor>`
- When dealing with a hierarchical DAG, with several segmentations, some intermediate tensors become I/O of sub-fusion
- When dealing with inter-device communication we need to pre-allocate recv buffers ; but those can be intermediate buffers
- In host pipelines, we sometimes need to pre-allocate the output and write to it slice by slice.

Reply:
FWIW, binding inputs and potential other preallocated buffers holds in one line in HostIrExecutor, :
https://github.com/NVIDIA/Fuser/blob/967c2824cfdf967901706d7aee1f57a9ab91f261/csrc/host_ir/executor.cpp#L69